### PR TITLE
Subtotal queries give incorrect results if query has a limit spec

### PR DESCRIPTION
**Pull request by loquisgon**
Description
For calculating the results for a subtotal spec, we reuse the buffer iterator that was used in the base query. However, not all the implementations of such iterators are reusable. Specifically, LimitedBufferHashGrouper#iterator is one of those non-reusable iterators. This results in subtotal specs not working if there is also a limit used as part of the query. 

This PR deeps copy some shared state in the iterator `makeHeapIteratormethod` to make sure that it is reusable. It also added a unit test to verify that it can be reused now.

To summarize:
* Issue is that `makeHeapIteratormethod` copies overwrite some shared state
* Solution in this ticket is to actually copy the state so that the iterator is reusable

This PR has:
- [x] been self-reviewed.
- [ ] added documentation for new or modified features or behaviors.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for code coverage is met.

### CODE REVIEW

File: sql/src/main/java/org/apache/druid/sql/SqlLifecycle.java [hunk 1]
```diff
@@ -268,7 +268,7 @@ public PlannerContext planAndAuthorize(final HttpServletRequest req)
     try {
       setParameters(SqlQuery.getParameterList(parameters));
       planAndAuthorize(authenticationResult);
-      result = execute();
+      result = execute(); //agt
```
Comment (loquisgon) [inline sql/src/main/java/org/apache/druid/sql/SqlLifecycle.java:271]: I will remove this debug comment

Comment (loquisgon) [reply sql/src/main/java/org/apache/druid/sql/SqlLifecycle.java]: Fixed...along with some style cleanup

File: processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/LimitedBufferHashGrouper.java [hunk 2]
```diff
@@ -318,6 +336,8 @@ public void close()
   private CloseableIterator<Entry<KeyType>> makeHeapIterator()
   {
     final int initialHeapSize = offsetHeap.getHeapSize();
+    ByteBufferMinMaxOffsetHeap newHeap = offsetHeap.copy();
```
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/LimitedBufferHashGrouper.java:339]: can you add a comment here on why we are creating another copy?

File: processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java [hunk 3]
```diff
@@ -59,6 +59,32 @@ public ByteBufferMinMaxOffsetHeap(
     this.heapIndexUpdater = heapIndexUpdater;
   }
 
+  public ByteBufferMinMaxOffsetHeap copy()
```
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java:62]: Let's also add some unit tests for this method.

File: processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/LimitedBufferHashGrouper.java [hunk 4]
```diff
@@ -318,6 +336,8 @@ public void close()
   private CloseableIterator<Entry<KeyType>> makeHeapIterator()
```
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/LimitedBufferHashGrouper.java:336]: Please add some unit tests in the `LimitedBufferHashGrouperTest` where an iterator is created multiple times and still returns the same results.

**Review by abhishekagarwal87 — COMMENTED** [review]
Looks mostly good except for unit tests.

File: processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/LimitedBufferHashGrouper.java (hunk 2 above)
Comment (loquisgon) [reply processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/LimitedBufferHashGrouper.java:339]: Done in next commit

File: processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java (hunk 3 above)
Comment (loquisgon) [reply processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java:62]: Done in next commit

File: processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/LimitedBufferHashGrouper.java (hunk 4 above)
Comment (loquisgon) [reply processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/LimitedBufferHashGrouper.java:336]: Done in next commit

File: processing/src/test/java/org/apache/druid/query/groupby/epinephelinae/LimitedBufferHashGrouperTest.java [hunk 5]
```diff
@@ -193,6 +194,65 @@ public void testMinBufferSize()
...
+      Assert.assertEquals(104, grouper.getSize());
+      Assert.assertEquals(209, grouper.getBuckets());
+      Assert.assertEquals(104, grouper.getMaxSize());
+    } else {
+      Assert.assertEquals(1899, grouper.getGrowthCount());
+      Assert.assertEquals(101, grouper.getSize());
+      Assert.assertEquals(202, grouper.getBuckets());
+      Assert.assertEquals(101, grouper.getMaxSize());
+    }
+    Assert.assertEquals(100, grouper.getLimit());
```
Comment (abhishekagarwal87) [inline processing/src/test/java/org/apache/druid/query/groupby/epinephelinae/LimitedBufferHashGrouperTest.java:243]:
```suggestion
  
```

Comment (abhishekagarwal87) [reply processing/src/test/java/org/apache/druid/query/groupby/epinephelinae/LimitedBufferHashGrouperTest.java:243]: these look unrelated to the unit test and are already being verified in a different unit test.

Comment (loquisgon) [reply processing/src/test/java/org/apache/druid/query/groupby/epinephelinae/LimitedBufferHashGrouperTest.java:243]: It does hurt though... i'll clean it up.

Comment (loquisgon) [reply processing/src/test/java/org/apache/druid/query/groupby/epinephelinae/LimitedBufferHashGrouperTest.java]: Done.

**Review by abhishekagarwal87 — APPROVED** [review]
LGTM. Thank you @loquisgon

File: processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java [hunk 6]
```diff
@@ -59,6 +60,35 @@ public ByteBufferMinMaxOffsetHeap(
...
+    int bufPosition = buf.position();
+    int bufLimit = buf.limit();
+    buf.rewind();
+
+    buffer.put(buf);
+    buffer.position(bufPosition);
+    buffer.limit(bufLimit);
+
+    buf.position(bufPosition);
+    buf.limit(bufLimit);
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java:83]: nit: It seems like this should be moved in to a separate function since it's used in 2 places. I tried a little googling, and didn't find any libraries that implement deep copying of ByteBuffers which I thought was surprising.

Looking at an implementation I found in http://www.java2s.com/example/java/java.nio/performs-a-deep-copy-on-a-byte-buffer.html, it looks like this implementation does not copy the order of the original ByteBuffer. I can't tell if this is important or not. What do you think?

File: processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java [hunk 7]
```diff
@@ -59,6 +60,35 @@ public ByteBufferMinMaxOffsetHeap(
...
+  public ByteBufferMinMaxOffsetHeap copy()
+  {
+    LimitedBufferHashGrouper.BufferGrouperOffsetHeapIndexUpdater updater =
+        Optional
+            .ofNullable(heapIndexUpdater)
+            .map(LimitedBufferHashGrouper.BufferGrouperOffsetHeapIndexUpdater::copy)
+            .orElse(null);
+
+    // deep copy buf
+    ByteBuffer buffer = ByteBuffer.allocateDirect(buf.capacity());
```
Comment (jihoonson) [inline processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java:72]: GroupBy queries use both processing buffers and merge buffers. The former is used when you compute per-segment results while the later is for all other purposes (merging per-segment results, computing subtotals, etc). Especially the broker only uses the merge buffers to process subtotals and subqueries. 

The merge buffers are maintained in a `BlockingPool` to manage the memory usage in brokers and historicals. This is important to avoid query failures due to OOM errors. Here, you should not allocate memory directly, but get one from the merge buffer pool. Check out [how the merge buffer is currently acquired](https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/GroupByRowProcessor.java#L117-L126).

Comment (loquisgon) [reply processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java:72]: Yeah, I agree. My biggest concern with this code is that it is not obvious to the caller that creating new iterators of the type being changed here will allocate new off-heap memory in an unbounded fashion. This is ok if we think that not "too many" copies will be done but I cannot affirm that.

File: processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java (hunk 6 above)
Comment (loquisgon) [reply processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java:83]: Yeah, agreed. Code should be refactored to abstract out the duplicated block.

Comment (loquisgon) [reply processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java:83]: It won't hurt to copy the order I guess. I will look more into it.

File: processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java (hunk 7 above)
Comment (jihoonson) [reply processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java:72]: > Yeah, I agree. My biggest concern with this code is that it is not obvious to the caller that creating new iterators of the type being changed here will allocate new off-heap memory in an unbounded fashion. This is ok if we think that not "too many" copies will be done but I cannot affirm that.

It seems pretty dangerous to me as users are not expected to be aware of this behavior. The blocking merge buffer pool is used to avoid unexpectedly using "too many" copies at the same time like in this problematic scenario. 

By the way, on the second look, I'm wondering why we copy the buffer instead of fixing the iterator of `LimitedBufferHashGrouper` to be re-iterable. This seems a better fix to me.

Comment (jihoonson) [reply processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java:72]: Talked with @loganlinn offline, we agreed fixing `LimitedBufferHashGrouper` would be a better fix.

**Comment by suneet-s** [comment]
Fixed by #10743
